Generals/Code/GameEngine/Source/GameClient/InGameUI.cpp

```diff
@@ -5428,21 +5428,33 @@ void InGameUI::selectNextIdleWorker( void )
 	else
 	{
 		Drawable *selectedDrawable = TheInGameUI->getFirstSelectedDrawable();	
+		ObjectList uniqueIdleWorkers;
+
+		// Find unique selectables to avoid selecting the same or a previous container if multiple idle workers are contained
+		for (ObjectListIt it = m_idleWorkers[index].begin(); it != m_idleWorkers[index].end(); ++it)
+		{
+			Object* itObj = *it;
+			while (itObj->getContainedBy())
+				itObj = itObj->getContainedBy(); // Treat containers as a single idle selectable
+
+			if (std::find(uniqueIdleWorkers.begin(), uniqueIdleWorkers.end(), itObj) == uniqueIdleWorkers.end())
+				uniqueIdleWorkers.push_back(itObj);
+		}
 		
-		ObjectListIt it = m_idleWorkers[index].begin();
-		while(it != m_idleWorkers[index].end())
+		ObjectListIt uit = uniqueIdleWorkers.begin();
+		while(uit != uniqueIdleWorkers.end())
 		{
-			Object *itObj = *it;
+			Object *itObj = *uit;
 			if(itObj == selectedDrawable->getObject())
 			{
-				++it;
-				if(it != m_idleWorkers[index].end())
-					selectThisObject = *it;
+				++uit;
+				if(uit != uniqueIdleWorkers.end())
+					selectThisObject = *uit;
 				else
-					selectThisObject = *m_idleWorkers[index].begin();
+					selectThisObject = *uniqueIdleWorkers.begin();
 				break;
 			}
-			++it;
+			++uit;
 		}
 		// if we had something selected that wasn't a worker, we'll get here
 		if(!selectThisObject)
```

Comment:
`*m_idleWorkers[index].begin()` here and further up needs looking at. It seems not right.

Reply:
Because it ignores the container bundling. So if the first Worker is in Container, then this behaves incorrectly. Also, code can be simplified to do `*m_idleWorkers[index].begin()` just once, instead of 2 places.

Reply:
What about the logic behaves incorrectly? The container bundling occurs further down. If the first worker is contained, subsequent workers are still selected as expected.

Reply:
First step:

```
selectThisObject = uniqueIdleWorkers.front();
```

down here.

At the top can do

```cpp
selectThisObject = *m_idleWorkers[index].begin();
while (selectThisObject->getContainedBy()) {
  selectThisObject = selectThisObject->getContainedBy();
```

Then 2x

```
while (itObj->getContainedBy())
  itObj = itObj->getContainedBy();
```

Can be simplified by adding a `getTopContainedBy()` function into Object class.

Reply:
Ok that does not help readability much. I initially thought we can get that down to one line.